Prefix Logs with Request ID

dyastremsky opened the pull request:
Prefix logs with the request ID, where possible.

Related changes:
Backend: https://github.com/triton-inference-server/backend/pull/60
Core: https://github.com/triton-inference-server/core/pull/85

rmccorm4 commented on src/grpc_server.cc:
```diff
@@ -3698,7 +3698,12 @@ ModelInferHandler::Process(InferHandler::State* state, bool rpc_ok)
     // COMPLETE. If error go immediately to COMPLETE.
     if (err != nullptr) {
-      LOG_VERBOSE(1) << "Infer failed: " << TRITONSERVER_ErrorMessage(err);
+      const char* request_id;
+      LOG_TRITONSERVER_ERROR(
+          TRITONSERVER_InferenceRequestIdString(irequest, &request_id),
+          "unable to retrieve request ID string");
+      LOG_VERBOSE(1) << request_id
+                     << "Infer failed: " << TRITONSERVER_ErrorMessage(err);
```
I'm concerned `request_id` may spit out some garbage data or have other side effects being uninitialized here if RequestIdString call fails in any of these places:
```cpp
rmccormick@RMCCORMICK:~/triton/jira/3780-trt-reload/tritonserver$ cat test2.cpp 
#include <iostream>

int main()
{
  const char* request_id;
  std::cout << "request_id: [" << request_id << "] and some text" << std::endl;

  const char* request_id_init = "";
  std::cout << "request_id_init: [" << request_id_init << "] and some text" << std::endl;
  return 0;
}
rmccormick@RMCCORMICK:~/triton/jira/3780-trt-reload/tritonserver$ g++ test2.cpp 
rmccormick@RMCCORMICK:~/triton/jira/3780-trt-reload/tritonserver$ ./a.out 
request_id: [] and some text
request_id_init: [] and some text
```

dyastremsky replied:
Good catch, and thanks for the code example. I initialized to an empty string to keep it simple.

The API call doesn't seem to have a way to fail currently (the error logging was to be extra safe), so there shouldn't be a case where request_id is overridden by garbage data.

GuanLuo commented on src/http_server.cc:
```diff
@@ -2737,7 +2737,12 @@ HTTPAPIServer::HandleInfer(
 
   if (err != nullptr) {
-    LOG_VERBOSE(1) << "Infer failed: " << TRITONSERVER_ErrorMessage(err);
+    const char* request_id = "";
```
Maybe the default can be "<id_unknown>"? If `TRITONSERVER_InferenceRequestId` ever returns error then we will print something instead of empty string

dyastremsky replied:
Thanks! Updated.

GuanLuo commented on src/http_server.cc:
```diff
@@ -2899,7 +2904,7 @@ HTTPAPIServer::InferRequestClass::FinalizeResponse(
       triton::common::TritonJson::ValueType::OBJECT);
 
-  const char* request_id;
+  const char* request_id = "";
```
Base on the check below (`request_id != nullptr`), should initialize to `nullptr`

dyastremsky replied:
Thanks, updated.

GuanLuo reviewed (approved):
small change needed, otherwise looks good to me

rmccorm4 commented on src/grpc_server.cc:
```diff
@@ -3679,6 +3679,14 @@ ModelInferHandler::Process(InferHandler::State* state, bool rpc_ok)
+    LOG_TRITONSERVER_ERROR(
+        TRITONSERVER_InferenceRequestId(irequest, &request_id),
+        "unable to retrieve request ID string");
+    if ((request_id != nullptr) && (request_id[0] != '\0')) {
```
do we need these checks for null terminator at start / empty string? Was there an edge case?

dyastremsky commented on qa/L0_lifecycle/lifecycle_test.py:
```diff
@@ -2568,7 +2568,8 @@ def callback(user_data, result, error):
         except InferenceServerException as ex:
             self.assertIn(
-                "in ensemble 'ensemble_zero_1_float32', Server is stopping, scheduler for model has stopped accepting new inference requests",
+                "in ensemble 'ensemble_zero_1_float32', [request id: <id_unknown>]" \
+                "Server is stopping, scheduler for model has stopped accepting new inference requests",
```
Sure, thanks for catching this early.

dyastremsky commented:
I rebased this ticket.
In addition, a few new commits:
- added checks for null or empty strings (default to unknown request ID)
- updated testing to expect the request_id
- moved the request_id getter to before inference to prevent a segfault when there is a request error (server does cleanup that results in a segfault when trying to call Id() after a failed inference, even if it's before we call delete).

This ticket is now passing CI (awaiting final run addressing Ryan's comment about the assert statement, but otherwise passes).

rmccorm4 commented on src/http_server.cc:
```diff
@@ -2910,7 +2910,7 @@ HTTPAPIServer::InferRequestClass::FinalizeResponse(
   const char* request_id = nullptr;
   RETURN_IF_ERR(TRITONSERVER_InferenceResponseId(response, &request_id));
-  if ((request_id != nullptr) && (request_id[0] != '\0')) {
+  if ((request_id == nullptr) || (request_id[0] == '\0')) {
```
I think this check should've remained `!=` in this spot, we don't need to add an empty or nullptr request_id to the response, right?

I would double check each of these request_id related if statements to see that they make sense in each spot in each of your PRs.

dyastremsky replied:
Thanks so much for your careful reviews. Everything should be good now.

rmccorm4 commented on src/http_server.cc:
```diff
@@ -2737,7 +2745,8 @@ HTTPAPIServer::HandleInfer(
 
   if (err != nullptr) {
-    LOG_VERBOSE(1) << "Infer failed: " << TRITONSERVER_ErrorMessage(err);
+    LOG_VERBOSE(1) << "[request id: " << request_id << "]"
```
Does it make more sense to use a common `GetRequestId()` for both servers in this PR too? See this comment: https://github.com/triton-inference-server/backend/pull/60/files#r912262759

rmccorm4 replied:
Same as other discussion, please add spaces after the "]" for consistency

rmccorm4 replied:
@dyastremsky

dyastremsky replied:
Sure, thanks. Done.

dyastremsky replied:
Sure, thanks. Done.
